man/man1/evolvotron.1
.TH EVOLVOTRON 1 "16 Oct 2009" "www.timday.com" "Evolvotron"

.SH NAME
evolvotron \- Creates generative art by an interactive evolutionary process.

.SH SYNOPSIS
evolvotron
[Qt options]
[options]

.SH DESCRIPTION

.B evolvotron
is interactive "generative art" software to evolve
images/textures/patterns through an iterative process of random
mutation and user-selection driven evolution.
Basically it displays a grid of random images; you click
on one you like and the rest of the grid is refilled with
variants of the one you picked.

This man page describes only the command line options.
A more complete manual fully describing usage via the application's GUI
is accessible from the application's "Help" menu once it is running.

.SH GENERAL OPTIONS

.TP 0.5i
.B \-a, \-\-autocool
Enable autocooling by default.

.TP 0.5i
.B  \-F, \-\-fullscreen
Start in "fullscreen" mode (window manager permitting).
[Press "Esc" key to revert to normal windowed mode].

.TP 0.5i
.B \-g, \-\-grid
.I colsxrows
Number of columns in image display grid (defaults to 5x6).

.TP 0.5i
.B \-h, \-\-help
Display a summary of command-line options and exit.

.TP 0.5i
.B \-j, \-\-jitter 
Enable sample jittering.

.TP 0.5i
.B \-m, \-\-multisample
.I multisample
Enables antialiased rendering.
Valid values are 1,2,3,4.
This specifies the size of the sub-pixel sampling grid,
so 1 provides the default one-sample-per-pixel behaviour,
while 4 provides 16 samples per pixel on a 4x4 grid.

.TP
.B \-M, \-\-menuhide
Start with menu and status bars suppressed.
[Press "Esc" key to display them].

.TP 0.5i
.B \-p, \-spheremap
Create spheremaps instead of planar textures.
NB The middle-mouse adjustments will not behave as expected in this mode.

.TP 0.5i
.B \-S, \-\-startup
.I filename
Specify a function file to be loaded on startup.
This option can be used multiple times, and any positional arguments will also be interpreted as such.
Loaded functions are placed on the grid from left to right, top to bottom; 
if multiple functions in excess of what is needed to fill the grid cells are provided, the extras will be ignored.
If the grid is not fully filled with startup functions, the remaining cells are filled at random normally.
Note that resetting the application will reload the specified function files again.
Also note that using this option multiple times has the potential to generate a lot of
loading error dialogs if e.g the path to all the files is wrong.

.TP 0.5i
.B \-U, \-\-shuffle
Use in conjunction with \-S / \-\-startup options, randomly shuffles the order the specified startup images are displayed in.
A fresh shuffle will be also be used when the application is reset.

.SH ANIMATION OPTIONS

.TP 0.5i
.B \-f, \-\-frames
.I frames
Number of frames to animate (defaults to 1 i.e no animaton)

.TP 0.5i
.B \-l, \-\-linear
Vary z linearly with time rather than sinusoidally over animation period.

.TP 0.5i
.B \-s, \-\-fps
.I framerate
Specify rate at which animations are displayed
(as an integer number of frames per second).
Defaults to 8.

.SH POWER-USER / DEBUG OPTIONS

.TP 0.5i
.B \-D, \-\-debug
Debug mode.
Currently simply sets function weightings so virtually all function nodes are FunctionNoiseOneChannel.
This is really only useful to developers in conjunction with the \-F/\-u options.

.TP 0.5i
.B \-E, \-\-enlarement-threadpool
Use a separate thread pool for computing enlargements.
Using this option ensures computation of enlargements continue to make
some progress even while the main grid is being actively worked on.
However, this will be at the expense of main grid rendering performance.
Without this option, enlargements' final high-resolution renderings are
invariably lower priority than computation for images in the main grid.
See also the \-N option to control the priority of threads in this pool.

.TP 0.5i
.B \-n, \-\-nice
.I niceness
Niceness of compute threads relative to the main application thread (defaults to 4).

.TP 0.5i
.B \-N, \-\-Nice
.I niceness
Niceness (relative to the main application thread) 
of compute threads dealing with enlargements (defaults to 8).
Only effective in conjunction with a separate enlargement threadool (\-\-E option).

.TP 0.5i
.I QtOptions
The Qt GUI system recognizes an number of additional options
(for example, standard X11 things like \-geometry <width>x<height>;
consult the Qt documentation for more information).
Note that these don't use the Gnu "double minus" option style
used for evolvotron options.

.TP 0.5i
.B \-t, \-\-threads
.I threads
Number of compute threads in a thread pool (defaults to number of CPUs)

.TP 0.5i
.B \-u, \-\-unwrapped
Use with the \-F option to stop the specified function from being wrapped by a random colouring and spatial transform node.

.TP 0.5i
.B \-v, \-\-verbose
Verbose mode.
Probably most useful for getting a list of supported
function names for use with the \-\-F option.

.TP 0.5i
.B \-x, \-\-favourite
.I functionname
Force a specific function type to be used at the top level of
all new image function trees (wrapped by random colour and
spatial transforms, unless the \-\-unwrapped option is also specified).
This can also be controlled from the "Favourite" dialog.

.SH EXAMPLES

evolvotron \-F FunctionSpiralLinear

evolvotron \-F FunctionKaleidoscope \-u

.SH AUTHOR
.B evolvotron
was written by Tim Day (www.timday.com) and is released
under the conditions of the GNU General Public License.
For further details see the application's "About" dialog
(accessible from the "Help" menu),
or the file LICENSE supplied with the source code.

.SH SEE ALSO

evolvotron_mutate(1), evolvotron_render(1)


NEWS
[Most recent at top]

Release 0.7.1:
 - Uses `CONFIG += ordered` in `main.pro`; enables parallelised make.
 - Fold in typos fixed by Debian downstream.
 - tarball on Sourceforge files.

Release 0.7.0:
 - Got to compile vs Qt5.9.1 on Debian Stretch & Jessie; preparing for Qt5-only on Buster.
 - Get rid of precompiled header use (more trouble than it's worth these days).
 - Upgrade std::auto_ptr use to std::unique_ptr - cleans up compiler's deprecated warning spew.

From release 0.6.3:
 - Version to 0.6.4

From release 0.6.2:
 - Version to 0.6.3
 - Files derived from USAGE (evolvotron.html and usage_text.h) are now under
   svn control and no longer need to be generated as part of the build of a
   fresh checkout.
 - Migrated to new SourceForge platform.
   Needed: svn switch --relocate https://evolvotron.svn.sourceforge.net/svnroot/evolvotron "svn+ssh://timday@svn.code.sf.net/p/evolvotron/code/"
   
From release 0.6.1:
 - Version to 0.6.2
 - CHANGES renamed to NEWS
 - License boilerplate managed using "headache"
 - Add some "tartan" inspired functions.
 - Add pixel/voxel/hex grid quantizing functions.
 - Port command line options to boost::program_options,
   rationalize commandline options.
 - Fix to evolvotron_mutate from Pedro Gimeno.
 - Update mkdeb script to work on squeeze
 - Functions to be loaded on startup can be specified on the commandline.
 - Add option to shuffle startup functions.
 
From release 0.6.0:
 - Version to 0.6.1
 - Fix bug in saved functions: quotes missing around version attribute

From release 0.5.1:
 - Version to 0.6.0
 - Switch to Qt4.
 - (Qt4 related) keyboard shortcuts rationalized
   many removed, ctrl-modifiers dropped
 - (Qt4 related) image displays now have a frame around them
 - (Qt4 related) Remove non-standard Qt animated PNG support
 - evolvotron_match removed
 - Platform specific code (e.g number of processors identification)
   isolated in libevolvotron/platform_specific.[h|cpp]
   and selected by PLATFORM_LINUX or PLATFORM_BSD
   (Fink Qt's __DARWIN_X11__ also detected and used to select PLATFORM_BSD).
 - Repository migrated to svn

From release 0.5.0:
 - Version to 0.5.1
 - Replace not often used frontside heating/cooling buttons with autocooling mechanism.
 - Separate tabs for info & license on "About" dialog.
 - Better formatting of function properties dialog, and also shows XML function tree.
 - Dialog with rendering controls
 - Multisampling (antialiasing) options
 - -a/-j/-s command line options for autocool/jitter/multisample
 - evolvotron_render adds -j/-s to control jitter & multisampling
 - Option (-E) of separate thread pool (with different nices: -n/-N) for enlargements.
   Task totals for each reported separately.
 - Enlargement and solo renderings are fragmented into multiple smaller
   jobs to better utilize multiprocessor systems.
 - Generally better use of ref-counted QImage and QPixmap.
 - Middle-mouse drag adjustments flicker less.

From release 0.4.0:
 - Version to 0.5.0
 - Lists of functions in alphabetic order.
 - New FunctionTop always at top level.
 - New functions: Frieze groups.
 - Constantness of functions more conservative (less overrides) and hopefully less buggy.
 - Final colour space transform range more constrained to keep colours sane.
 - More sensible mutation of top level spatial transform (less arbitrary shearing).
 - Note that above changes mean old saved functions will appear different.
 - New mutation parameter: probability of a function's parameters being reset.
 - Code clean up:
   - More std::auto_ptr, boost::shared_ptr, boost::ptr_vector.  Code is now delete free.
   - Get rid of hacked static function registration; uses auto-generated register_all_functions.
 - Compute threads use QWaitCondition, not polling.
 - Default number of threads is number of processors.
 - Compute threads run at lower priority.
 - -n <niceness> command line option to control compute thread priority.
   (NB assumes Linux NPTL's non-Posix compliant behaviour of threads NOT sharing a common nice value;
   see "man pthreads" on Linux.)
 - About dialog displays number of threads.
 - Use precompiled headers (build time 2m45s down to 1m45s).
 - code_review script
 - Set an icon on various dialogs and main frame.
 - Add -D (function debug) massively weighting in favour of FunctionNoiseOneChannel.  Useful with -x/-X.
 - Believed to fix Debian bug #436231 and gentoo #144160
 - MutatableImage has serial number (fix problems caused by pointer value being wrongly used as equivalent)

From release 0.3.1:
 - Version to 0.4.0
 - Lock state moved to mutatable image so is stored in undo history
 - Mandelbrot & Julia set only in x,y plane.  Julia set uses fixed parameters.
 - Add Juliabrot.
 - Fix "Randomize" button on "All" functions tab.
 - "Favourite" function dialog provides access to -x/-X command-line functionality
 - Undo after reset works better
 - Fix broken FunctionGradient.
   Replace with GradientOfMagnitude, DivergenceOfMagnitude, Divergence, Curl
 - Migrate float->double via typedef real
 - Always builds with fullscreen option.

From release 0.3.0:
 - Version to 0.3.1
 - Fix for gcc 3.4
 - ./configure made (slightly) more robust

From release 0.2.3:
 - Version to 0.3.0
 - Get rid of troublesome system call to ../VERSION 
   in common.pro and pass settings from configure script instead
 - Add not-very-useful .qt-mng image file output format
 - Add function stats to Properties dialog
 - Add function simplification (more useful for debugging than anything)
 - Don't use True and False in doc-building python script
   (seems to be quite new thing, not in 2.1.x ?).  Use 0 and 1 instead.
 - Internal changes
 - Add pkg/debian .deb builder script (still under development)
 - Fix buggy Y rotation matrix
 - Eliminate template-tastic Tuple and Matrix classes.
   Fix XYZ and Transform to not require them.
 - evolvotron_mutate - -g option more likely to generate insteresting functions
 - -linz and -spheremap options where appropriate

evolvotron/evolvotron.cpp
/**************************************************************************/
/*  Copyright 2012 Tim Day                                                */
/*                                                                        */
/*  This file is part of Evolvotron                                       */
/*                                                                        */
/*  Evolvotron is free software: you can redistribute it and/or modify    */
/*  it under the terms of the GNU General Public License as published by  */
/*  the Free Software Foundation, either version 3 of the License, or     */
/*  (at your option) any later version.                                   */
/*                                                                        */
/*  Evolvotron is distributed in the hope that it will be useful,         */
/*  but WITHOUT ANY WARRANTY; without even the implied warranty of        */
/*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         */
/*  GNU General Public License for more details.                          */
/*                                                                        */
/*  You should have received a copy of the GNU General Public License     */
/*  along with Evolvotron.  If not, see <http://www.gnu.org/licenses/>.   */
/**************************************************************************/

/*! \file
  \brief Application code for evolvotron executable (contains "main").
*/

/*! \mainpage Evolvotron : image evolver

  \author Tim Day 

  \section introduction Introduction
  "Evolvotron" is an interactive tool for producing "generative art".
  Images are generated from function trees, which are then mutated and evolved through a process of user selection.

  \todo For new features to be added, see the TODO file.
 */

#include "evolvotron_main.h"
#include "platform_specific.h"

#include <boost/program_options.hpp>

//! Application code
int main(int argc,char* argv[])
{
  QApplication app(argc,argv);
  
  // General options
  bool autocool;
  bool fullscreen;
  std::string grid;
  bool help;
  bool jitter;
  bool menuhide;
  uint multisample;
  bool spheremap;
  std::vector<std::string> startup;
  bool startup_shuffle;
  boost::program_options::options_description options_desc("General options");
  {
    using namespace boost::program_options;
    options_desc.add_options()
      ("autocool,a"   ,bool_switch(&autocool)                         ,"Enable autocooling")
      ("fullscreen,F" ,bool_switch(&fullscreen)                       ,"Fullscreen window")
      ("grid,g"       ,value<std::string>(&grid)->default_value("6x5"),"Columns x rows in image grid")
      ("help,h"       ,bool_switch(&help)                             ,"Print command-line options help message and exit")
      ("jitter,j"     ,bool_switch(&jitter)                           ,"Enable rendering jitter")
      ("multisample,m",value<uint>(&multisample)->default_value(1)    ,"Multisampling grid (NxN)")
      ("menuhide,M"   ,bool_switch(&menuhide)                         ,"Hide menus")
      ("spheremap,p"  ,bool_switch(&spheremap)                        ,"Generate spheremaps")
      ("startup,S"    ,value<std::vector<std::string> >(&startup)     ,"Startup function (multiples allowed, or positional)")
      ("shuffle,U"    ,bool_switch(&startup_shuffle)                  ,"Shuffle startup functions")
      ;
  }

  // Animation options
  int framerate;
  int frames;
  bool linear;
  boost::program_options::options_description animation_options_desc("Animation options");
  {
    using namespace boost::program_options;
    animation_options_desc.add_options()
      ("frames,f"     ,value<int>(&frames)->default_value(1)      ,"Frames in an animation")
      ("linear,l"     ,bool_switch(&linear)                       ,"Sweep z linearly in animations")
      ("fps,s"        ,value<int>(&framerate)->default_value(8)   ,"Animation speed (frames-per-second)")
      ;
  }

  // Advanced options
  bool debug;
  bool enlargement_threadpool;
  std::string favourite;
  int niceness_enlargement;
  int niceness_grid;
  uint threads;
  bool unwrapped;
  bool verbose;

  boost::program_options::options_description advanced_options_desc("Advanced options");
  {
    using namespace boost::program_options;
    advanced_options_desc.add_options()
      ("debug,D"                 ,bool_switch(&debug)                    ,"Enable function debug mode")
      ("enlargement-threadpool,E",bool_switch(&enlargement_threadpool)   ,"Enlargements computed using a separate threadpool")
      ("nice,n"                  ,value<int>(&niceness_grid)->default_value(4)
       ,"Niceness of compute threads for image grid")
      ("Nice,N"                  ,value<int>(&niceness_enlargement)->default_value(8)
       ,"Niceness of compute threads for enlargements (if separate pool)")
      ("threads,t"               ,value<uint>(&threads)->default_value(get_number_of_processors())
       ,"Number of threads in a thread pool")
      ("unwrapped,u"             ,bool_switch(&unwrapped)                ,"Don't wrap favourite function")
      ("verbose,v"               ,bool_switch(&verbose)                  ,"Log some details to stderr")
      ("favourite,x"             ,value<std::string>(&favourite)         ,"Favourite function")
      ;
  }

  boost::program_options::positional_options_description positional_options_desc;
  positional_options_desc.add("startup",-1);

  options_desc.add(animation_options_desc);
  options_desc.add(advanced_options_desc);
  
  boost::program_options::variables_map options;
  boost::program_options::store(
    boost::program_options::command_line_parser(
      argc,argv
    ).options(
      options_desc
    ).positional(
      positional_options_desc
    ).run()
    ,options
  );
  boost::program_options::notify(options);

  if (help)
    {
      std::cerr << options_desc;
      return 0;
    }
  
  if (verbose)
    std::clog.rdbuf(std::cerr.rdbuf());
  else
    std::clog.rdbuf(sink_ostream.rdbuf());

  //! \todo Could be done better maybe (set 'x' as input separator)
  const std::string::size_type p=grid.find("x");
  if (p==std::string::npos || p==0 || p==grid.size()-1)
    {
      std::cerr << "--grid option argument isn't in <rows>x<cols> format\n";
      return 1;
    }
  else
    {
      grid[p]=' ';
    }
  int cols=6;
  int rows=5;
  std::stringstream(grid) >> cols >> rows;

  if (cols*rows<2)
    {
      std::cerr << "Must be at least 2 display grid cells\n";
      return 1;
    }

  if (frames<1)
    {
      std::cerr << "Must specify at least 1 frame\n";
      return 1;
    }

  if (framerate<1)
    {
      std::cerr << "Must specify framerate of at least 1\n";
      return 1;
    }

  std::clog
    << "Evolvotron version "
    << stringify(EVOLVOTRON_BUILD)
    << " starting with " 
    << cols 
    << " by " 
    << rows 
    << " display cells and " 
    << threads
    << " compute threads per farm (niceness "
    << niceness_grid
    << " and "
    << niceness_enlargement
    << ")\n";

  if (!startup.empty()) {
    std::clog << "Startup functions to be loaded: ";
    for (size_t i=0;i<startup.size();++i) {
      if (i>0) std::clog << ", ";
      std::clog << startup[i];
    }
    std::clog << "\n";
  }

  EvolvotronMain*const main_widget=new EvolvotronMain
      (
       0,
       QSize(cols,rows),
       frames,
       framerate,
       threads,
       enlargement_threadpool,
       niceness_grid,
       niceness_enlargement,
       fullscreen,
       menuhide,
       autocool,
       jitter,
       multisample,
       debug,
       linear,
       spheremap,
       startup,
       startup_shuffle
       );

  main_widget->mutation_parameters().function_registry().status(std::clog);

  if (!favourite.empty())
    {
      std::clog
	<< "Selected specific function: "
	<< favourite
	<< (unwrapped ? " (unwrapped)" : " (wrapped)")
	<< "\n";

      if (!main_widget->favourite_function(favourite))
	{
	  std::cerr << "Unrecognised function name specified for -x/-X option\n";
	  return 1;
	}

      main_widget->favourite_function_unwrapped(unwrapped);
    }
  
  main_widget->show();
  
  // NB Do this here rather than in constructor so that compute threads aren't being fired off during general GUI set-up
  
  std::clog << "Resetting main widget...\n";
  main_widget->reset_cold();
  
  // NB No need to worry about deleting EvolvotronMain... QApplication seems to do it for us.
  std::clog << "Commencing main loop...\n";
  return app.exec();
  std::clog << "...returned from main loop\n";
  
  return 0;
}
